In [1]:
import json 

with open('emp.json', 'r') as fin:
    employees = json.load(fin)


In [2]:
import pymongo # pymongo is a python driver for MongoDB
import credentials # load username and password from credentials.py
connection_string = "mongodb+srv://chamarthyyeswanth:6A7170d0HXrAGO3h@cluster01.4ihbtj6.mongodb.net/"

In [3]:
client = pymongo.MongoClient(connection_string)

In [4]:
db = client['employees']

In [5]:
if 'emp_01' in db.list_collection_names():
    db['emp_01'].drop()

In [6]:
for emp in employees['employees']:
    db['emp_01'].insert_one(emp)

 I have loaded the employee data which contains information about employee name , department, salary and position. I have synthesized and prepared the data which is generated by chatgpt in emp.json file. Data represents employee details within an organization.

id : Unique identifier for the employee.

name : Name of the employee.

position :  job position or title of the employee within the organization.

department : Department to which the employee belongs. This could be areas like Engineering, Marketing, Product Management, Sales, Finance, Design, Customer Service

salary :  Annual compensation of the employee. The salary amount can vary for each employee.

In [7]:
emp_collection = db['emp_01']

In [8]:
averages = emp_collection.aggregate([
    {
        "$match": {
            "department": {"$ne": None}
        }
    },
    {
        "$group": {
            "_id": "$department",
            "avg_salary": {"$avg": "$salary"}
        }
    },
    {
        "$sort": {
            "avg_salary": -1
        }
    }
])

x =list(averages)
print(x)

[{'_id': 'Product Management', 'avg_salary': 95000.0}, {'_id': 'Engineering', 'avg_salary': 86250.0}, {'_id': 'Design', 'avg_salary': 85000.0}, {'_id': 'Marketing', 'avg_salary': 85000.0}, {'_id': 'Sales', 'avg_salary': 75000.0}, {'_id': 'Finance', 'avg_salary': 70000.0}, {'_id': 'Customer Service', 'avg_salary': 55000.0}]


I'm attempting to gather insights from the employee data. It is important for me to find out the average salary for employees in each department within the organization.

$match Stage: This stage filters out employee records with a "department" field other than None. By doing this, we ensure that only valid department entries are considered in the aggregation.

$group Stage: Using the "_id" key, the data is then grouped by the "department" field in the $group stage.The $avg accumulator is used to calculate the average salary within each of the groups.

$sort Stage: Finally, in the $sort stage, the results are sorted using the "avg_salary" field in descending order. By sorting the departments by average salary in descending order, the highest salaries are displayed at the top.


Results from the query provide valuable insight into the organization's compensation structure. It shows that the highest average salary is $150,000 in the Product Management department, followed by $86,250 in Engineering, $85,000 in Design, $85,000 in Marketing, $75,000 in Sales, and $70,000 in Finance.


In [9]:


with open("avgemp.json", "w") as fout:
    json.dump(x, fout, indent=4)